File: libraries/ios/Sources/VoidhashCore/Network/NetworkPolicy.swift
import Foundation

/// The shared transport policy: what is worth retrying, how long to wait and how to read
/// `Retry-After`.
///
/// Every queue, refresh and outbox in the SDK uses these values so behaviour under an outage is
/// identical no matter which subsystem issued the request.
public enum NetworkPolicy {
    /// Per-request timeout applied to every SDK `URLSession`.
    public static let requestTimeoutSeconds: TimeInterval = 10
    /// Whole-resource timeout applied to every SDK `URLSession`.
    public static let resourceTimeoutSeconds: TimeInterval = 30
    /// Backoff ceiling for queue and outbox retries.
    public static let queueBackoffCapMilliseconds: Double = 30_000
    /// Backoff ceiling for background config refreshes.
    public static let configBackoffCapMilliseconds: Double = 60_000
    /// Longest an interactive read waits on an in-flight refresh before serving a stale value.
    public static let freshnessBudgetMilliseconds: Double = 500

    /// Statuses that mean "try again later" rather than "this request was wrong".
    public static let retryableStatusCodes: Set<Int> = [408, 429, 500, 502, 503, 504]

    private static let breakerFailureStatusCodes: Set<Int> = [408, 500, 502, 503, 504]

    /// Whether a response indicates host unavailability rather than a client verdict or throttle.
    public static func countsTowardsCircuitBreaker(statusCode: Int?) -> Bool {
        guard let statusCode else {
            return true
        }
        return breakerFailureStatusCodes.contains(statusCode)
    }

    /// Whether `statusCode` should be retried. A transport failure (no status) always is.
    public static func isRetryable(statusCode: Int?) -> Bool {
        guard let statusCode else {
            return true
        }
        return retryableStatusCodes.contains(statusCode)
    }

    /// Whether `statusCode` means the publishable key was rejected.
    public static func isAuthFailure(statusCode: Int?) -> Bool {
        return statusCode == 401 || statusCode == 403
    }

    /// Jittered exponential backoff: `min(cap, 1s · 2^(attempt-1))` plus up to 25 % jitter.
    ///
    /// - Parameters:
    ///   - attempt: 1 for the first retry.
    ///   - cap: Ceiling before jitter.
    ///   - jitter: Fraction of the base delay added on top, injectable for deterministic tests.
    public static func backoffMilliseconds(
        attempt: Int,
        cap: Double = queueBackoffCapMilliseconds,
        jitter: Double = Double.random(in: 0..<0.25)
    ) -> Double {
        let exponent = Double(max(attempt - 1, 0))
        // pow overflows to infinity well before the cap matters; clamp the exponent first.
        let base = min(1000 * pow(2, min(exponent, 32)), cap)
        return base + base * max(jitter, 0)
    }

    /// Parses `Retry-After`, accepting a delay in seconds and an HTTP date, header first and the
    /// body's `retry_after_ms` second.
    ///
    /// - Parameters:
    ///   - header: Raw `Retry-After` header value.
    ///   - body: Response body, searched for `retry_after_ms` when the header is absent.
    ///   - now: Millisecond epoch used to turn an HTTP date into a delay.
    ///   - cap: Ceiling applied to the parsed delay, so a server asking for an hour cannot park a
    ///     queue past the backoff ceiling its owner chose. `nil` returns the raw value.
    public static func retryAfterMilliseconds(
        header: String?,
        body: Data? = nil,
        now: Double,
        cap: Double? = nil
    ) -> Double? {
        return clampRetryAfter(rawRetryAfterMilliseconds(header: header, body: body, now: now), cap: cap)
    }

    /// Clamps an already parsed `Retry-After` delay to `cap`; `nil` passes through.
    public static func clampRetryAfter(_ milliseconds: Double?, cap: Double?) -> Double? {
        guard let milliseconds else {
            return nil
        }
        guard let cap else {
            return milliseconds
        }
        return min(milliseconds, cap)
    }

    private static func rawRetryAfterMilliseconds(
        header: String?, body: Data?, now: Double
    ) -> Double? {
        if let fromHeader = parseRetryAfterHeader(header, now: now) {
            return fromHeader
        }
        guard let body, !body.isEmpty,
            let object = try? JSONSerialization.jsonObject(with: body) as? [String: Any]
        else {
            return nil
        }
        guard let milliseconds = (object["retry_after_ms"] as? NSNumber)?.doubleValue,
            milliseconds >= 0
        else {
            return nil
        }
        return milliseconds
    }

    private static func parseRetryAfterHeader(_ value: String?, now: Double) -> Double? {
        guard let value else {
            return nil
        }
        let trimmed = value.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else {
            return nil
        }

        if let seconds = Double(trimmed), seconds >= 0 {
            return (seconds * 1000).rounded(.up)
        }

        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "GMT")
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        guard let date = formatter.date(from: trimmed) else {
            return nil
        }
        return max(date.timeIntervalSince1970 * 1000 - now, 0)
    }

    /// Process-wide session carrying the SDK timeouts, used wherever `.shared` used to be.
    public static let defaultSession: URLSession = makeSession()

    /// Builds a `URLSession` carrying the SDK timeouts.
    public static func makeSession() -> URLSession {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = requestTimeoutSeconds
        configuration.timeoutIntervalForResource = resourceTimeoutSeconds
        return URLSession(configuration: configuration)
    }
}

File: libraries/android/sdk/src/main/java/com/voidhash/sdk/network/NetworkPolicy.kt
package com.voidhash.sdk.network

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.ceil
import kotlin.random.Random

/** Statuses the SDK retries. Everything else is a verdict the server will repeat. */
val RETRYABLE_STATUS_CODES: Set<Int> = setOf(408, 429, 500, 502, 503, 504)

private val BREAKER_FAILURE_STATUS_CODES: Set<Int> = setOf(408, 500, 502, 503, 504)

/** Whether a response indicates host unavailability rather than a client verdict or throttle. */
fun countsTowardsCircuitBreaker(status: Int): Boolean = status in BREAKER_FAILURE_STATUS_CODES

/** Backoff ceiling for queue flushes. */
const val QUEUE_BACKOFF_CAP_MS: Long = 30_000L

/** Backoff ceiling for configuration refreshes (schema, person, flags, paywalls). */
const val CONFIG_BACKOFF_CAP_MS: Long = 60_000L

/** How long an interactive read waits on an in-flight refresh before serving a stale value. */
const val FRESHNESS_BUDGET_MS: Long = 500L

private const val BACKOFF_BASE_MS = 1_000L

private val HTTP_DATE_FORMATS = listOf(
    "EEE, dd MMM yyyy HH:mm:ss zzz",
    "EEE, dd-MMM-yy HH:mm:ss zzz",
    "EEE MMM d HH:mm:ss yyyy",
)

/**
 * Jittered exponential backoff: `min(cap, 1 s · 2^(attempt - 1)) + rand(0, 25 %)`.
 *
 * @param attempt 1-based attempt number.
 * @param capMs upper bound before jitter.
 */
fun backoffDelayMs(
    attempt: Int,
    capMs: Long = QUEUE_BACKOFF_CAP_MS,
    random: Random = Random.Default,
): Long {
    val exponent = (attempt - 1).coerceIn(0, 30)
    val base = minOf(BACKOFF_BASE_MS shl exponent, capMs)
    return base + random.nextLong((base / 4).coerceAtLeast(1L))
}

/**
 * Parses a `Retry-After` cool-down, header first and response body second.
 *
 * The header carries either a delay in seconds or an HTTP date; the body carries
 * `retry_after_ms`. Returns `null` when neither is present or parseable. [capMs] clamps the
 * result to the caller's backoff ceiling: a server asking for an hour must not park a queue
 * for longer than the SDK would ever back off on its own.
 */
fun parseRetryAfterMs(
    header: String?,
    body: String? = null,
    now: Long = 0L,
    capMs: Long? = null,
): Long? {
    val parsed = parseRetryAfterHeaderMs(header, now) ?: parseRetryAfterBodyMs(body) ?: return null
    return if (capMs == null) parsed else parsed.coerceAtMost(capMs)
}

private fun parseRetryAfterHeaderMs(header: String?, now: Long): Long? {
    val raw = header?.trim().orEmpty()
    if (raw.isEmpty()) return null

    raw.toDoubleOrNull()?.let { seconds ->
        if (seconds < 0) return null
        return ceil(seconds * 1000).toLong()
    }

    for (pattern in HTTP_DATE_FORMATS) {
        val parsed = runCatching {
            SimpleDateFormat(pattern, Locale.US)
                .apply { timeZone = TimeZone.getTimeZone("UTC") }
                .parse(raw)
        }.getOrNull() ?: continue
        // An HTTP-date in the past means "retry now", not "retry in the past".
        return (parsed.time - now).coerceAtLeast(0L)
    }
    return null
}

private fun parseRetryAfterBodyMs(body: String?): Long? {
    if (body.isNullOrBlank()) return null
    val json = runCatching { org.json.JSONObject(body) }.getOrNull() ?: return null
    if (!json.has("retry_after_ms") || json.isNull("retry_after_ms")) return null
    val milliseconds = (json.opt("retry_after_ms") as? Number)?.toDouble() ?: return null
    return milliseconds.takeIf { it.isFinite() && it >= 0 }?.let { ceil(it).toLong() }
}

/** Formats [epochMillis] as the ISO-8601 instant the ingest API expects. */
internal fun formatIsoTimestamp(epochMillis: Long): String {
    val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
    formatter.timeZone = TimeZone.getTimeZone("UTC")
    return formatter.format(Date(epochMillis))
}
